api/model/src/main/java/org/projectnessie/model/Util.java

```diff
@@ -25,45 +25,231 @@
+              elements.add(e.toString());
+              e.setLength(0);
+            }
+          } else {
+            c = encoded.charAt(++i);
+            switch (c) {
+              case ESCAPE_FOR_DOT:
+                e.append(DOT);
+                break;
+              case ESCAPE_FOR_SLASH:
+                e.append(SLASH);
+                break;
+              case ESCAPE_FOR_BACKSLASH:
+                e.append(BACKSLASH);
+                break;
+              case ESCAPE_FOR_PERCENT:
+                e.append(PERCENT);
+                break;
+              case DOT:
+                elements.add(e.toString());
+                e.setLength(0);
+                break;
+              default:
+                // Any other character, that character is the first character of the _next_ element;
+                // throw new IllegalArgumentException("Illegal escape sequence character '"+c+" at
+                // index : " + c);
+                --i;
+                elements.add(e.toString());
+                e.setLength(0);
+                break;
+            }
+          }
+          break;
+        case GROUP_SEPARATOR:
+        case ZERO_BYTE:
+          e.append(DOT);
+          break;
+        default:
+          e.append(c);
+          break;
+      }
+    }
+    if (e.length() > 0) {
+      elements.add(e.toString());
+    }
+    return elements;
+  }
+
+  /** Escapes content-key elements, see {@link Elements#toPathString()}. */
+  public static String toPathString(List<String> elements) {
+    StringBuilder sb = new StringBuilder();
+    for (String element : elements) {
+      if (sb.length() > 0) {
+        sb.append('.');
+      }
+      int l = element.length();
+      for (int i = 0; i < l; i++) {
+        char c = element.charAt(i);
+        sb.append(c == DOT || c == ZERO_BYTE ? GROUP_SEPARATOR : c);
+      }
+    }
+    return sb.toString();
+  }
+
+  /** Escapes content-key elements, see {@link Elements#toPathStringEscaped()}. */
+  public static String toPathStringEscaped(List<String> elements) {
+    StringBuilder sb = new StringBuilder();
+    boolean escaped = false;
+    for (String element : elements) {
+      if (sb.length() > 0) {
+        sb.append(DOT);
+        if (escaped) {
+          sb.append(DOT);
+        }
+      }
+      escaped = escapePathElement(element, sb, escaped);
+    }
+    return sb.toString();
+  }
+
+  /** Escapes all characters, that are "problematic" in URI paths. */
+  static boolean escapePathElement(String element, StringBuilder sb, boolean escaped) {
+    int sbl = sb.length();
+    int l = element.length();
+    for (int i = 0; i < l; i++) {
+      char c = element.charAt(i);
+      switch (c) {
+        case DOT:
+        case SLASH:
+        case BACKSLASH:
+        case PERCENT:
+          if (!escaped) {
+            // Not yet "escaped", let the first escaped element start with a '.'
+            if (sb.length() > sbl) {
+              sb.insert(sbl, DOT);
```

Comment:
WDYT about putting the escaping mark (`.`) always as the first char of the whole string? I think it would make writing escaped strings easier for humans.

For example: `foo..._bar` becomes `.foo.._bar` - it it easy to see that escaping is "on" and it is easier to interpret the second dot group as a separator followed by an escaped char.

I still feel that using a different rare char for escaping, makes the string even more human-readable IMHO, e.g. if `*` is the escape char, `foo..._bar` becomes `.foo.~.bar`, `.foo._bar` -> `.foo~.bar`, `.foo~bar` -> `.foo~~bar`.

Futhermore, in escaped mode the escape char is always followed by one of the special chars (`~.` - dot `.{` - slash, etc..). `.` is _always_ the elements separator.

This should also help with canonicalization. We could say that canonical strings must be in escape more. Still, simply adding a leading `.` will convert the key to its canonical form, e.g. `foo.bar` -> `.foo.bar`. Of course, `~` will have to be escaped: `foo~bar` -> `.foo~~bar`, `foo/bar` -> `.foo~{bar`. However, my point is that it is fairly intuitive for humans to do.

Also, there is no ambiguity in constructing the canonical string, while currently `foo.._bar` and `foo._bar` both parse to `["foo", "_bar"]`.

Both unescaped and canonical strings will be accepted on input, but we will be able to use only canonical strings in CEL expressions, hashing, and other places that require deterministic behaviour.

While `~` is converted to `%7E` by `URLEncoder`, it is "unreserved" in [RFC 3986](https://www.rfc-editor.org/rfc/rfc3986), so humans will be able to use it with `curt` directly, e.g. `curl http://localhost:19120/api/v2/trees/main/contents/.foo.~.bar`

More  examples: (array, handwritten, canonical, for URI path)

`["a","b","c"]` - `a.b.c` - `.a.b.c` -  (either handwritten or canonical accepted)
`["a","b~c"]` - `a.b~c` - `.a.b~~c` - (either handwritten or canonical accepted)
`["a","b.c"]` - `a.b/c` - `.a.b/c` - `.a.b~{c`
`["a","b.c"]` - `.a.b~.c` - `.a.b~.c` - `.a.b~.c`

WDYT?